src/System.Management.Automation/engine/Subsystem/SubsystemManager.cs

```diff
@@ -0,0 +1,254 @@
+            {
+                return subsystemInfo;
+            }
+
+            throw new ArgumentException(
+                StringUtil.Format(
+                    SubsystemStrings.SubsystemTypeUnknown,
+                    subsystemType.FullName));
+        }
+
+        /// <summary>
+        /// Get the information about a subsystem by the subsystem kind.
+        /// </summary>
+        /// <param name="kind">A specific <see cref="SubsystemKind"/>.</param>
+        /// <returns>The <see cref="SubsystemInfo"/> object that represents the concrete subsystem.</returns>
+        public static SubsystemInfo GetSubsystemInfo(SubsystemKind kind)
+        {
+            if (s_subSystemKindMap.TryGetValue(kind, out SubsystemInfo subsystemInfo))
+            {
+                return subsystemInfo;
+            }
+
+            throw new ArgumentException(
+                StringUtil.Format(
+                    SubsystemStrings.SubsystemKindUnknown,
+                    kind.ToString()));
+        }
+
+        #endregion
+
+        #region public - Subsystem registration
+
+        /// <summary>
+        /// Subsystem registration.
+        /// </summary>
+        /// <typeparam name="TConcreteSubsystem">The concrete subsystem base type.</typeparam>
+        /// <typeparam name="TImplementation">The implementation type of that concrete subsystem.</typeparam>
+        /// <param name="proxy">An instance of the implementation.</param>
+        public static void RegisterSubsystem<TConcreteSubsystem, TImplementation>(TImplementation proxy)
+            where TConcreteSubsystem : class, ISubsystem
+            where TImplementation : class, TConcreteSubsystem
+        {
+            if (proxy == null)
+            {
+                throw new ArgumentNullException(nameof(proxy));
+            }
+
+            RegisterSubsystem(GetSubsystemInfo(typeof(TConcreteSubsystem)), proxy);
+        }
+
+        /// <summary>
+        /// Register an implementation for a subsystem.
+        /// </summary>
+        /// <param name="kind">The target <see cref="SubsystemKind"/> of the registration.</param>
+        /// <param name="proxy">An instance of the implementation.</param>
+        public static void RegisterSubsystem(SubsystemKind kind, ISubsystem proxy)
+        {
+            if (proxy == null)
+            {
+                throw new ArgumentNullException(nameof(proxy));
+            }
+
+            if (kind != proxy.Kind)
+            {
+                throw new ArgumentException(
+                    StringUtil.Format(
+                        SubsystemStrings.ImplementationMismatch,
+                        proxy.Kind.ToString(),
+                        kind.ToString()),
+                    nameof(proxy));
+            }
+
+            RegisterSubsystem(GetSubsystemInfo(kind), proxy);
+        }
+
+        private static void RegisterSubsystem(SubsystemInfo subsystemInfo, ISubsystem proxy)
```

Comment:
Is subsystem registration persisted, say in a configuration file?

Reply:
It's not by this API, but the pre-registration of selected subsystems should be supported eventually to allow shipping customized PowerShell package that contains the selected subsystems out-of-box.
Configuration could be a way, another option is to define a specific folder in `$PSHOME` for the core engine to discover the available subsystems.